tinygrad/tensor.py

```diff
@@ -724,22 +724,22 @@ def _walk(node, visited):
         yield node
     return list(_walk(self, set()))
 
-  def backward(self) -> Tensor:
+  def backward(self, gradient:Optional[Tensor]=None) -> Tensor:
     """
     Propagates the gradient of a tensor backwards through the computation graph.
-    Must be used on a scalar tensor.
+    Must be called on a scalar tensor or the `gradient` argument must be provided.
 
     ```python exec="true" source="above" session="tensor" result="python"
     t = Tensor([1.0, 2.0, 3.0, 4.0], requires_grad=True)
     t.sum().backward()
     print(t.grad.numpy())
     ```
     """
-    assert self.shape == tuple(), f"backward can only be called for scalar tensors, but it has shape {self.shape})"
-
+    assert self.shape == tuple() or gradient is not None, "backward must be called on a scalar tensor or with the gradient argument"
+    assert gradient is None or gradient.shape == self.shape, f"gradient shape must match tensor shape, {gradient.shape!r} != {self.shape!r}"
     # fill in the first grad with one. don't use Tensor.ones because we don't need contiguous
     # this is "implicit gradient creation"
-    self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False)
+    self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False) if gradient is None else gradient
```

Comment:
Create the implicit gradient before the asserts

Reply:
you mean something like
```python
self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False)
assert self.shape == tuple() or gradient is not None, "backward must be called on a scalar tensor or with the gradient argument"
assert gradient is None or gradient.shape == self.shape, f"gradient shape must match tensor shape, {gradient.shape!r} != {self.shape!r}"
self.grad = self.grad if gradient is None else gradient
```
?

Reply:
but this could be incorrect. 'gradient' (which is really a vector jacobian product) must be of the same shape as the tensor (self.shape == self.grad.shape) always. So creating an 'implicit' (scalar with value 1) grad (vector-jacobian product) for a tensor of different shape is incorrect. Did you mean something else?

Reply:
No. The same code.

```
if gradient is None:
  assert self.shape == tuple()
  gradient = ...```

Reply:
> No. The same function, don't change self.grad
> 
> ```
> if gradient is None:
>   assert self.shape == tuple()
>   gradient = ...```
> ```

Corrected in the last commit